### PR TITLE
Add support for replication

### PR DESCRIPTION
**Pull request by vkalintiris**
##### Summary

This PR adds support for replicating data through the streaming protocol:

- The parent maintains a list of gaps.
- Whenever a new agent is connected, the parent creates a new gap based on the connection time and the last entry saved in dbengine for the child host.
- On connection the parent sends the list of gaps to the child.
- A _replication_ thread on the child is responsible for collecting, processing and sending the data of every gap received by the parent. This is done through the streaming protocol by adding two new commands:
  - FILLGAP \<base64-encoded-protobuf-message\> (contains chart, dimension, timestamps & stored values)
  - DROPGAP \<after> \<before>

For each gap, a child will send a FILLGAP command for all the dimensions that are active. Once all dimensions have been processed, a DROPGAP command is sent to the parent to let it know that the child won't send any more data for that gap.

##### Test Plan

- Start the child agent and let it run for a couple minutes. Start the parent and after a while the child host on the parent will have charts without gaps. Try all the variations of this testing, eg. kill/start the parent (a) immediately, (b) after a while, (c) repeatedly, etc.

- For active-active parents with two children, the following test was done (`c1`: child 1, `c2`: child 2, `p1`: parent1, `p2`: parent2):

```
t1:  start c1
t2:  start c2
t3:  start p1
t4:  start p2
t5:  kill  p1
t6:  start p1
t7:  kill  p1
t8:  kill  p2
t9:  start p1
t10: start p2
```

- To enable replication you can add this to your `netdata.conf`:
```
[replication]
   enabled = yes
   log replication operations = yes
```
By enabling logging you can visit the `host/<hostname>/api/v1/replication` endpoint and watch the full history of the replication events.

### CODE REVIEW

File: streaming/replication/Utils.h [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+#include "replication-private.h"
+
+namespace replication {
+
+/*
+ * Wraps a standard std::mutex and enables/disables cancelability
+ * in lock/unlock operations.
+ */
+class Mutex {
```
Comment (ktsaou) [inline streaming/replication/Utils.h:12]: Is this object needed?
The standard mutex functions of netdata, already do this.

File: streaming/replication/Utils.h [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        netdata_thread_enable_cancelability();
+        M.unlock();
+    }
+
+    bool try_lock() {
+        netdata_thread_disable_cancelability();
+        if (M.try_lock())
+            return true;
+
+        netdata_thread_disable_cancelability();
```
Comment (ktsaou) [inline streaming/replication/Utils.h:29]: should be enable, not disable.

**Comment by ktsaou** [comment]
@vkalintiris I think that netdata library functions should always be used when they are available:

* threads management should always be done with the `netdata_*` functions
* locks and mutexes management should always be done with the `netdata_*` functions
* memory allocations should always be done with the netdata `z` functions
* error logging
etc

Generally, if netdata already provides functions for something, we should use it in place of anything else. If something is missing in them, we should add it. If something is wrong with them we should fix it.

Failure to do so, means that the code will become increasingly more difficult to manage, understand and debug and we will lose the ability to improve the project as a whole by just improving a component of it.

I also grab the opportunity to discuss the general use of C++:

Generally, I am ok in using C++ when required, but we should agree on a policy about it. I believe C++ (or any other language) should be used to the absolute minimum required. So, it is ok to write in C++ a small component that is difficult to implement in C and is ready to use in C++, but it is not ok to switch something to C++ just for convenience, which would otherwise be almost the same effort to write in C, with some C++ assistance.

Don't get me wrong. I strongly believe that object oriented programming techniques, like well defined constructors and destructors on all objects, use of private and public methods and attributes for isolation between modules, etc should be always be used as programming best-practices, but this does not necessarily mean switching trivial code to C++.

Let's discuss this.

Looking forward to see this merged!

PS: btw, I found this repo that discusses and benchmarks all base64 implementations: https://github.com/gaspardpetit/base64 have a look. It also mentions the version you added.

File: streaming/replication/Utils.h (hunk 1 above)
Comment (vkalintiris) [reply streaming/replication/Utils.h:12]: Resolving this because it belongs in the greater C vs. C++ comment you made and we need to have that discussion.

Comment (vkalintiris) [reply streaming/replication/Utils.h:12]: FWIW, it's needed to enable scope-based locks that automatically lock/unlock mutexes based on scope.

**Comment by ktsaou** [comment]
@vkalintiris what is the status of this?

**Comment by vkalintiris** [comment]
> @vkalintiris what is the status of this?

I'm running tests to check every corner case I can think of, with the underlying dbengine + query APIs. There are a couple of issues I've found, eg:

```
writing data in [-128-512, -128], ie. within the active page
internal query api returns the correct values:
   3 2022-06-16 16:32:50: netdata ERROR : WEB_SERVER[static1] : GVD: timestamp=1655383688, value=412.000000
   2 2022-06-16 16:32:50: netdata ERROR : WEB_SERVER[static1] : GVD: timestamp=1655383816, value=62.000000         <--------- OK (verified by printing the value at rrdset_done)
   1 2022-06-16 16:32:50: netdata ERROR : WEB_SERVER[static1] : GVD: timestamp=1655383817, value=113.000000
however, high level query (grouping) returns:
 [ 1655383817,113],           <--------- OK
 [ 1655383816,null],            <--------- why?!
...
...
...
 [ 1655383692,null],
 [ 1655383691,null],
 [ 1655383690,null],
 [ 1655383689,62],              <--------- why?!
 [ 1655383688,412],           <--------- OK
Query URL:
http://nx570:19999/api/v1/data?chart=prof_type.profchart_0&format=json&options=jsonwrap%7Cnonzero&after=1655383378&before=1655383817
```

I need to keep testing this today so I'll have more to share on Monday.

**Comment by ktsaou** [comment]
rebase please... there are a lot that have been fixed...

**Comment by vkalintiris** [comment]
> @vkalintiris what is the status of this?

I've run a couple tests for active-active parents with two children. Overall, there are two blockers:

1. The replication thread pushes data to a parent for a dimension that does not have an active descriptor yet. Usually, this happens for dimensions with an `update_every > 1 second` when the replication thread sends data before the streaming thread. The proper fix requires us to add some logic to create an active descriptor when one is not found. The easy way to bypass this is to wait for the maximum update every of all dimensions in the host.
2. In an active-active parent configuration, when the active parent is SIGINT'd the failover parent has a gap in between the time it takes for the children to handle the disconnection from the active parent and switch to the failover one. This is a timing issue that I need to investigate and fix.

Other than these two main issues, the replication seems to work well for long duration of gaps, ie. I managed to backfill data spanning multiple hours between active parent -> failover parent, and children -> active parent.

There are also two non-blocking issues (at least from my perspective):

1. The issue described at https://github.com/netdata/netdata/pull/13048#issuecomment-1158605556 (which I have not checked again yet, after rebasing my branch).
2. When a child is connected to a parent, the value of a single timestamp is duplicated twice in the parent. For example, if the values of a dimensions are: `t(i) = x`, `t(i+1) = y` and `t(i+2) = z` then the parent will present the dimension's values as `t(i) = x`,  `t(i+1) = x` and `t(i+2) = z`.

**Comment by thiagoftsm** [comment]
@vkalintiris , ṕlease, update `plugins.d` documentation with the two new keywords.

File: database/engine/rrdenginerepl.c [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+    page = (storage_number *)descr->pg_cache_descr->page;
+    page_gap = (storage_number *)dim_past_data->page;
+
+    uint64_t start, end, page_start, page_end;
+    start = dim_past_data->start_time / USEC_PER_SEC; //gap time start
+    end = dim_past_data->end_time / USEC_PER_SEC;     //gap time end
+    page_start = descr->start_time / USEC_PER_SEC;    //active page time start
+    page_end = descr->end_time / USEC_PER_SEC;        //active page time start
+
+    if (!page || !page_gap || start > end) {
```
Comment (thiagoftsm) [inline database/engine/rrdenginerepl.c:186]: You can do these tests before all the divisions that happens between lines `180-184`.

File: database/sqlite/sqlite_functions.c [hunk 4]
```diff
@@ -2254,3 +2263,216 @@ void sql_load_node_id(RRDHOST *host)
...
+    param++;
+    rc = sqlite3_bind_int(res, 3, Before);
+    if (unlikely(rc != SQLITE_OK))
+        goto bind_fail;
+
+    rc = execute_insert(res);
+    if (unlikely(rc != SQLITE_DONE))
+        error_report("Failed to insert host entries range, rc = %d", rc);
+
+    error("GVD: saved host entries range to sqlite: <%ld, %ld>", After, Before);
```
Comment (thiagoftsm) [inline database/sqlite/sqlite_functions.c:2309]: Why are you reporting an `error` here? Is it not better to have an `info`?

File: database/engine/rrdenginerepl.c (hunk 3 above)
Comment (vkalintiris) [reply database/engine/rrdenginerepl.c:186]: Not really because the error message shows the computed information.

File: database/sqlite/sqlite_functions.c (hunk 4 above)
Comment (vkalintiris) [reply database/sqlite/sqlite_functions.c]: Left over from when I was testing this manually. I've removed these log messages because they don't help at all to understand what's going on in the replication component. For debugging potential problems it's better to enable `log replication operations` and watch the output of the `api/v1/replication` endpoint.

**Comment by stelfrag** [comment]
Notice a core dump but unfortunately not much info there -- just

```
#90 0x000055aaae342b2c in send_gaps_data (rpt=<error reading variable: Cannot access memory at address 0x100000000ffdfe8>) at streaming/receiver.c:429
```

File: database/sqlite/sqlite_functions.c [hunk 5]
```diff
@@ -2313,3 +2322,210 @@ void sql_load_node_id(RRDHOST *host)
 
     return;
 };
+
+#define SQL_REPLICATION_SAVE_HOST_ENTRIES_RANGE \
+    "INSERT OR REPLACE INTO replication_host_entries_range(host_id, after, before) VALUES (?1, ?2, ?3);"
+
+int replication_save_host_entries_range(uuid_t *host_id, time_t after, time_t before)
+{
+    sqlite3_stmt *res = NULL;
```
Comment (stelfrag) [inline database/sqlite/sqlite_functions.c:2331]: This need to be `static __thread sqlite3_stmt *res = NULL`

File: database/sqlite/sqlite_functions.c [hunk 6]
```diff
@@ -2313,3 +2322,210 @@ void sql_load_node_id(RRDHOST *host)
...
+        error_report("Failed to reset statement in replication_save_host_entries_range function, rc = %d", rc);
+    return rc;
+}
+
+#define SQL_REPLICATION_LOAD_HOST_ENTRIES_RANGE \
+    "SELECT after, before FROM replication_host_entries_range WHERE host_id == ?1;"
+
+int replication_load_host_entries_range(uuid_t *host_id, time_t *after, time_t *before)
+{
+    sqlite3_stmt *res = NULL;
```
Comment (stelfrag) [inline database/sqlite/sqlite_functions.c:2387]: This need to be static __thread sqlite3_stmt *res = NULL

File: database/sqlite/sqlite_functions.c [hunk 7]
```diff
@@ -2313,3 +2322,210 @@ void sql_load_node_id(RRDHOST *host)
...
+        error_report("Failed to reset statement in replication_load_host_entries_range function, rc = %d", rc);
+    return rc;
+}
+
+#define SQL_REPLICATION_SAVE_GAPS \
+    "INSERT OR REPLACE INTO replication_gaps(host_id, gaps) VALUES (?1, ?2);"
+
+int replication_save_gaps(uuid_t *host_id, const char *Buf, size_t Len)
+{
+    sqlite3_stmt *res = NULL;
```
Comment (stelfrag) [inline database/sqlite/sqlite_functions.c:2435]: This need to be static __thread sqlite3_stmt *res = NULL

**Comment by vkalintiris** [comment]
@stelfrag rebased + 2 new commits.

**Comment by stelfrag** [comment]
```
#0  __memmove_avx_unaligned () at ../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S:222
#1  0x00007fb0fab1e6d8 in std::basic_streambuf<char, std::char_traits<char> >::xsputn(char const*, long) () from /lib/x86_64-linux-gnu/libstdc++.so.6
#2  0x00007fb0fab10b45 in std::basic_ostream<char, std::char_traits<char> >& std::__ostream_insert<char, std::char_traits<char> >(std::basic_ostream<char, std::char_traits<char> >&, char const*, long) () from /lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x000055a7ce13d417 in replication::GapData::print (this=0x7faf984df200, RH=0x55a87de39d00) at streaming/replication/GapData.cc:13
#4  0x000055a7ce13e56b in replication::GapData::flushToDBEngine (this=0x7faf984df200, RH=0x55a87de39d00) at streaming/replication/GapData.cc:180
```

**Comment by vkalintiris** [comment]
@stelfrag It seems that we can hit this issue frequently. Give it one more try, now that errors are printed in separate lines, to get the precise variable causing the error.

**Comment by stelfrag** [comment]
```
#0  __memmove_avx_unaligned_erms () at ../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S:548
#1  0x00005598021dcf4b in rrdeng_store_past_metrics_page (dim_past_data=0x7fe6cf1de110) at database/engine/rrdenginerepl.c:72
#2  0x00005598021f4f57 in replication::GapData::flushToDBEngine (this=0x7fe6cf1df1e0, RH=0x559809c0f460) at streaming/replication/GapData.cc:211
#3  0x00005598021fdac7 in replication_receiver_fill_gap (RH=0x559809c0f460, 
    Buf=0x55980ce510a9 "ChZzeXN0ZW0ucHJvY2Vzc2VzX3N0YXRlEgdydW5uaW5nGoUoworUlgbCitSWBsKK1JYGworUlgbCitSWBsKK1JYGworUlgbCitSWBsKK1JYGworUlgbCitSWBsKK1JYGworUlgbCitSWBsKK1JYGworUlgbCitSWBsKK1JYGworUlgbCitSWBsKK1JYGworUlgbCitSW"...) at streaming/replication/replication.cc:427
#4  0x0000559802097299 in pluginsd_fillgap_action (user=0x7fe6cf1e3440, host=0x559809c0f460, 
    buf=0x55980ce510a9 "ChZzeXN0ZW0ucHJvY2Vzc2VzX3N0YXRlEgdydW5uaW5nGoUoworUlgbCitSWBsKK1JYGworUlgbCitSWBsKK1JYGworUlgbCitSWBsKK1JYGworUlgbCitSWBsKK1JYGworUlgbCitSWBsKK1JYGworUlgbCitSWBsKK1JYGworUlgbCitSWBsKK1JYGworUlgbCitSW"...) at collectors/plugins.d/pluginsd_parser.c:212
#5  0x0000559802098e56 in pluginsd_fillgap (words=0x7fe6cf1df330, user=0x7fe6cf1e3440, plugins_action=0x559809a90520) at collectors/plugins.d/pluginsd_parser.c:660
#6  0x0000559802225e65 in parser_action (parser=0x559809b12000, input=0x55980ce510a0 "FILLGAP") at parser/parser.c:296
#7  0x00005598021f0353 in streaming_parser (rpt=0x55980ce51000, cd=0x7fe6cf1e35a0, fp=0x559809593ac0) at streaming/receiver.c:406
#8  0x00005598021f2175 in rrdpush_receive (rpt=0x55980ce51000) at streaming/receiver.c:790
#9  0x00005598021f24e9 in rrdpush_receiver_thread (ptr=0x55980ce51000) at streaming/receiver.c:837
#10 0x0000559801e4428a in thread_start (ptr=0x5598061caff0) at libnetdata/threads/threads.c:185
#11 0x00007fe71d11bb43 in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:442
#12 0x00007fe71d1ada00 in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81

```

File: streaming/replication/replication.cc [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+    snprintfz(LogPath, FILENAME_MAX, "%s/%s.fg",
+              netdata_configured_cache_dir, rrdhost_hostname(RH));
+    FILE *FP = fopen(LogPath, "a");
+    if (!FP) {
+        error("Could not open log file: %s", LogPath);
+        return false;
+    }
+    size_t BufLen = strlen(Buf);
+    fprintf(FP, "%zu %s\n", BufLen, Buf);
+    fclose(FP);
```
Comment (vkalintiris) [inline streaming/replication/replication.cc:420]: @stelfrag This piece of code will log every FILLGAP command received by upstream to `var/netdata/cache/<hostname>.fg`.

If you get a crash within `pluginsd_fillgap_action`, I can check if it's caused by a serialization/deserialization error.

I didn't manage to make the child/parent agents crash during my testing. However, after rebasing on top of master, I'm noticing that not all gaps are filled completely (which should be a separate issue from the crashes that we are trying to investigate now).

**Comment by vkalintiris** [comment]
Ignore the updates in this PR. It has been rebased on top of master and I want to check the build failures we'll get from the various distros we are supporting.

The commits have been squashed because changes are required all over the agent. I'll document the changes in each component of the replication's architecture in the PR's description along with the testing that we need to perform in order to verify that we didn't break anything.

**Comment by thiagoftsm** [comment]
@vkalintiris, please, take a look in [this](https://github.com/netdata/netdata/pull/13048#issuecomment-1257571044) LGTM report.
